services/src/main/java/org/keycloak/headers/DefaultSecurityHeadersProvider.java

```diff
@@ -0,0 +1,168 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.headers;
+
+import org.jboss.logging.Logger;
+import org.keycloak.models.BrowserSecurityHeaders;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+
+import javax.ws.rs.InternalServerErrorException;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerResponseContext;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import java.util.Map;
+
+public class DefaultSecurityHeadersProvider implements SecurityHeadersProvider {
+
+    private static final Logger LOGGER = Logger.getLogger(DefaultSecurityHeadersProvider.class);
+
+    private final Map<String, String> headerValues;
+    private final KeycloakSession session;
+
+    private DefaultSecurityHeadersOptions options;
+
+    public DefaultSecurityHeadersProvider(KeycloakSession session) {
+        this.session = session;
+
+        RealmModel realm = session.getContext().getRealm();
+        if (realm != null) {
+            headerValues = realm.getBrowserSecurityHeaders();
+        } else {
+            headerValues = BrowserSecurityHeaders.defaultHeaders;
+        }
+    }
+
+    @Override
+    public SecurityHeadersOptions options() {
+        if (options == null) {
+            options = new DefaultSecurityHeadersOptions();
+        }
+        return options;
+    }
+
+    @Override
+    public void addHeaders(ContainerRequestContext requestContext, ContainerResponseContext responseContext) {
+        if (options != null && options.isSkipHeaders()) {
+            return;
+        }
+
+        MediaType requestType = requestContext.getMediaType();
+        MediaType responseType = responseContext.getMediaType();
+        MultivaluedMap<String, Object> headers = responseContext.getHeaders();
+
+        if (responseType == null && !isEmptyMediaTypeAllowed(requestContext, responseContext)) {
+            LOGGER.errorv("MediaType not set on path {0}, with response status {1}", session.getContext().getUri().getRequestUri().getPath(), responseContext.getStatus());
+            throw new InternalServerErrorException();
+        }
+
+        if (isRest(requestType, responseType)) {
+            addRestHeaders(headers);
+        } else if (isHtml(requestType, responseType)) {
+            addHtmlHeaders(headers);
+        } else {
+            addGenericHeaders(headers);
+        }
+    }
+
+    private void addGenericHeaders(MultivaluedMap<String, Object> headers) {
```

Comment:
Not sure if it will be beneficial to have this method "protected" to allow easier overriding of this method in case that someone wants to have custom provider? Same question for "addRestHeaders", "addHtmlHeaders", "addHeader" .

Reply:
Don't like opening up internal classes too much. Someone can also extend be calling the default provider within their extended provider, to then modify the results. It's more likely that folks will want to change one or two headers and not implement all the headers for one "type".

Reply:
Sure, this is very minor thing. Just wanted to doublecheck.